### DISCUSSION TITLE
Is the Mutex necessary?

### DISCUSSION THREAD

**Discussion by fn-jt** [orbit-mvi/orbit-mvi#14] [Q&A] [answered]
Hi! Thanks for open-sourcing this slick library. I've been studying different implementations of "state containers", and I'm trying to get a handle around the sorts of concurrency bugs that could arise from them. I noticed that in the implementation of RealContainer, `internalStateFlow.value` is updated inside of a lock. Is this necessary, or just an extra precaution you guys took? I've seen other implementations just do something like:
```.kt
        scope.launch {
            while (isActive)
                store.value = intents.receive().reduce(store.value)
        }
```
Where intents is a Channel<Intent>. The implementation I'm referencing is here: https://github.com/kanawish/upvote/blob/flow/app/src/main/java/com/kanawish/upvote/common/FlowModelStore.kt.
Is there some sort of concurrency bug this implementation would fall prey to?

**Comment by fn-jt** [comment]
Actually, I think I understand what's going on after reading this: https://kotlinlang.org/docs/reference/coroutines/shared-mutable-state-and-concurrency.html. You guys chose to use a Mutex instead of thread confinement while the other implementation chose (UI) thread confinement.

**Comment by Rosomack** [comment, marked as answer]
Exactly! Originally we used a `newSingleThreadContext` to confine reductions to. However switching to that thread in coroutines is considered expensive. Using a mutex instead allows us to avoid expensive context switching. Additionally,  a `Mutex` in coroutines is much more light-weight than Java's mutexes, as they are **suspending** rather than blocking.

The kotlin team is [going to deprecate this `newSingleThreadContext`](https://github.com/Kotlin/kotlinx.coroutines/issues?q=is%3Aissue+is%3Aopen+sort%3Aupdated-desc+pool) in favour of creating concurrency limited dispatchers backed by the default dispatcher.
